Mathlib/InformationTheory/Hamming.lean

```diff
@@ -199,214 +208,189 @@ end HammingDistNorm
+  of (i : ι) : β i
 
 namespace Hamming
 
-variable {α ι : Type*} {β : ι → Type*}
-
-/-! Instances inherited from normal Pi types. -/
+variable {α ι : Type*} {β γ : ι → Type*}
 
-instance [∀ i, Inhabited (β i)] : Inhabited (Hamming β) :=
-  ⟨fun _ => default⟩
-
-instance [DecidableEq ι] [Fintype ι] [∀ i, Fintype (β i)] : Fintype (Hamming β) :=
-  Pi.instFintype
-
-instance [Inhabited ι] [∀ i, Nonempty (β i)] [Nontrivial (β default)] : Nontrivial (Hamming β) :=
-  Pi.nontrivial
+/-- `Hamming.ofEquiv` is the equivalence between `Hamming` and the corresponding Pi type. -/
+@[simps]
+def ofEquiv : Hamming β ≃ ∀ i, β i where
+  toFun := of
+  invFun := mk
 
-instance [Fintype ι] [∀ i, DecidableEq (β i)] : DecidableEq (Hamming β) :=
-  Fintype.decidablePiFintype
+@[simp] theorem of_inj {a b : Hamming β} : a.of = b.of ↔ a = b := ofEquiv.injective.eq_iff
 
-instance [∀ i, Zero (β i)] : Zero (Hamming β) :=
-  Pi.instZero
+@[ext] protected theorem ext {a b : Hamming β} (h : a.of = b.of) : a = b := ofEquiv.injective h
 
-instance [∀ i, Neg (β i)] : Neg (Hamming β) :=
-  Pi.instNeg
+/-- The coordinate-wise map between Hamming types. -/
+@[simps]
+def map (f : ∀ i, γ i → β i) (x : Hamming γ) : Hamming β where
+  of := Pi.map f x.of
 
-instance [∀ i, Add (β i)] : Add (Hamming β) :=
-  Pi.instAdd
-
-instance [∀ i, Sub (β i)] : Sub (Hamming β) :=
-  Pi.instSub
-
-instance [∀ i, SMul α (β i)] : SMul α (Hamming β) :=
-  Pi.instSMul
-
-instance [Zero α] [∀ i, Zero (β i)] [∀ i, SMulWithZero α (β i)] : SMulWithZero α (Hamming β) :=
-  Pi.smulWithZero _
-
-instance [∀ i, AddMonoid (β i)] : AddMonoid (Hamming β) :=
-  Pi.addMonoid
-
-instance [∀ i, AddGroup (β i)] : AddGroup (Hamming β) := Pi.addGroup
+/-! Instances inherited from normal Pi types. -/
 
-instance [∀ i, AddCommMonoid (β i)] : AddCommMonoid (Hamming β) :=
-  Pi.addCommMonoid
+instance [∀ i, Inhabited (β i)] : Inhabited (Hamming β) := ofEquiv.inhabited
 
-instance [∀ i, AddCommGroup (β i)] : AddCommGroup (Hamming β) :=
-  Pi.addCommGroup
+instance [DecidableEq ι] [Fintype ι] [∀ i, Fintype (β i)] : Fintype (Hamming β) :=
+  Fintype.ofEquiv _ ofEquiv.symm
 
-instance (α) [Semiring α] (β : ι → Type*) [∀ i, AddCommMonoid (β i)] [∀ i, Module α (β i)] :
-    Module α (Hamming β) :=
-  Pi.module _ _ _
+instance [Inhabited ι] [∀ i, Nonempty (β i)] [Nontrivial (β default)] : Nontrivial (Hamming β) :=
+  ofEquiv.nontrivial
 
-/-! API to/from the type synonym. -/
+instance [Fintype ι] [∀ i, DecidableEq (β i)] : DecidableEq (Hamming β) := ofEquiv.decidableEq
 
+instance [∀ i, Zero (β i)] : Zero (Hamming β) := ofEquiv.zero
 
-/-- `Hamming.toHamming` is the identity function to the `Hamming` of a type. -/
-@[match_pattern]
-def toHamming : (∀ i, β i) ≃ Hamming β :=
-  Equiv.refl _
+instance [∀ i, Neg (β i)] : Neg (Hamming β) := ofEquiv.Neg
 
-/-- `Hamming.ofHamming` is the identity function from the `Hamming` of a type. -/
-@[match_pattern]
-def ofHamming : Hamming β ≃ ∀ i, β i :=
-  Equiv.refl _
```

Comment:
Why not keep these names as the constructor / projection?

Reply:
I think they are unwieldy: I was inspired here by `[Matrix.of](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Matrix/Defs.html#Matrix.of)`. `Hamming.ofHamming` to me is just obviously less ergonomic than `Hamming.of`.

Reply:
But `Matrix.of` is to `Matrix` as your Hamming.mk` is to `Hamming`.

Reply:
Oh. Yes! How silly of me. I still think `toHamming` and `ofHamming` are not ideal names, though.